src/QueryBuilder.ts

```diff
@@ -151,12 +173,42 @@ export abstract class QueryBuilder<T> {
 
         this.validate();
 
-        const response = await client._unaryCall(node.url, this._inner, this._method);
+        await setTimeoutAwaitable(receiptInitialDelayMs);
 
-        const responseHeader = getResponseHeader(response);
-        throwIfExceptional(responseHeader.getNodetransactionprecheckcode());
+        /* eslint-disable no-await-in-loop */
+        // we want to wait in a loop, that's the whole point here
+        for (let attempt = 0; /* loop will exit when transaction expires */; attempt += 1) {
+            const response = await client._unaryCall(node.url, this._inner, this._method);
 
-        return this._mapResponse(response);
+            const responseHeader = getResponseHeader(response);
+            let status: number = responseHeader.getNodetransactionprecheckcode();
+
+            if (response.getResponseCase() === ResponseCase.TRANSACTIONGETRECEIPT) {
+                status = response.getTransactiongetreceipt()!.getReceipt()!.getStatus();
+            }
+
+            // typecast required or we get a mismatching union type error
+            if (([
+                ResponseCodeEnum.BUSY,
+                ResponseCodeEnum.UNKNOWN,
+                ResponseCodeEnum.OK
+            ] as number[])
+                .includes(status)) {
+                const delay = Math.floor(receiptRetryDelayMs *
+                    Math.random() * ((2 ** attempt) - 1));
+
+                if (Date.now() + delay > validUntilMs) {
+                    throw new Error(`timed out waiting for consensus on transaction ID: ${txId}`);
```

Comment:
Just throw `HederaError` with the last status. This will be talking about `transaction ID: null` any query but receipts anyway.

Reply:
It'd be a little confusing if the user got `HederaError(OK)` but we're already doing that in the Java SDK, plus the fewer error types the user has to think about the better.